src/services/code-index/embedders/jina.ts

```diff
@@ -0,0 +1,278 @@
+import { IEmbedder, EmbeddingResponse, EmbedderInfo } from "../interfaces"
+import { getModelQueryPrefix } from "../../../shared/embeddingModels"
+import { t } from "../../../i18n"
+import {
+	withValidationErrorHandling,
+	formatEmbeddingError,
+	getErrorMessageForStatus,
+} from "../shared/validation-helpers"
+import type { HttpError } from "../shared/validation-helpers"
+import { TelemetryEventName } from "@roo-code/types"
+import { TelemetryService } from "@roo-code/telemetry"
+import {
+	MAX_BATCH_TOKENS,
+	MAX_ITEM_TOKENS,
+	MAX_BATCH_RETRIES as MAX_RETRIES,
+	INITIAL_RETRY_DELAY_MS as INITIAL_DELAY_MS,
+} from "../constants"
+
+interface JinaEmbeddingRequest {
+	model: string
+	input: string[]
+	encoding_type?: "float" | "base64"
+	task?: string
+	dimensions?: number
+	late_chunking?: boolean
+	embedding_type?: "float" | "base64" | "binary" | "ubinary"
+}
+
+interface JinaEmbeddingResponse {
+	model: string
+	object: "list"
+	usage: {
+		total_tokens: number
+		prompt_tokens: number
+	}
+	data: Array<{
+		object: "embedding"
+		index: number
+		embedding: number[] | string
+	}>
+}
+
+/**
+ * Jina implementation of the embedder interface with batching and rate limiting
+ * Uses jina-embeddings-v4 with multi-vector embeddings for code search
+ */
+export class JinaEmbedder implements IEmbedder {
+	private readonly apiKey: string
+	private readonly baseUrl = "https://api.jina.ai/v1"
+	private readonly defaultModelId: string
+
+	/**
+	 * Creates a new Jina embedder
+	 * @param apiKey Jina API key
+	 * @param modelId Optional model identifier (defaults to jina-embeddings-v4)
+	 */
+	constructor(apiKey: string, modelId?: string) {
+		this.apiKey = apiKey
+		this.defaultModelId = modelId || "jina-embeddings-v4"
```

Comment:
Consider adding model validation against the known supported models (jina-embeddings-v4, jina-embeddings-v3, jina-clip-v2) for better error messages. You could add a validation check in the constructor:

```suggestion
const SUPPORTED_MODELS = ['jina-embeddings-v4', 'jina-embeddings-v3', 'jina-clip-v2'];
if (modelId && !SUPPORTED_MODELS.includes(modelId)) {
    console.warn(`Unknown Jina model: ${modelId}. Supported models: ${SUPPORTED_MODELS.join(', ')}`);
}
```